File: matmul/matmul.py
# Writen by Attractive Chaos; distributed under the MIT license

import array
import platform
import socket
import sys
import os


def matmul(a, b):
    c = [array.array("d", [0.0]) * len(b) for _ in range(len(b[0]))]
    for i in range(len(b[0])):
        for j in range(len(b)):
            c[i][j] = b[j][i]

    d = [array.array("d", [0.0]) * len(b[0]) for _ in range(len(a))]

    for i in range(len(a)):
        for j in range(len(b[0])):
            s, ai, cj = 0.0, a[i], c[j]
            for k in range(len(b)):
                s += ai[k] * cj[k]
            d[i][j] = s
    return d


def build_matrix(n, seed):
    t = seed / n / n
    m = [array.array("d", [0.0]) * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            m[i][j] = t * (i - j) * (i + j)
    return m


def calc(n):
    n = n // 2 * 2
    a = build_matrix(n, 1.0)
    b = build_matrix(n, 2.0)

    d = matmul(a, b)
    return d[n // 2][n // 2]


def notify(msg):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        if not s.connect_ex(("localhost", 9001)):
            s.sendall(bytes(msg, "utf8"))


if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100

    left = calc(101)
    right = -18.67
    if abs(left - right) > 0.1:
        print("%f != %f" % (left, right), file=sys.stderr)
        quit(1)

    notify("%s\t%d" % (platform.python_implementation(), os.getpid()))
    results = calc(n)
    notify("stop")

    print(results)

File: matmul/matmul.c
// Writen by Attractive Chaos; distributed under the MIT license

#include <libnotify.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#ifdef __clang__
#define COMPILER "clang"
#else
#define COMPILER "gcc"
#endif

double **mm_init(int n) {
  double **m = (double **)malloc(n * sizeof(void *));
  for (int i = 0; i < n; ++i) {
    m[i] = calloc(n, sizeof(double));
  }
  return m;
}

void mm_destroy(int n, double **m) {
  for (int i = 0; i < n; ++i) {
    free(m[i]);
  }
  free(m);
}

double **mm_gen(int n, double seed) {
  double tmp = seed / n / n;
  double **m = mm_init(n);
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      m[i][j] = tmp * (i - j) * (i + j);
    }
  }
  return m;
}

// better cache performance by transposing the second matrix
double **mm_mul(int n, double *const *a, double *const *b) {
  double **m = mm_init(n);
  double **c = mm_init(n);
  for (int i = 0; i < n; ++i) { // transpose
    for (int j = 0; j < n; ++j) {
      c[i][j] = b[j][i];
    }
  }
  for (int i = 0; i < n; ++i) {
    double *p = a[i], *q = m[i];
    for (int j = 0; j < n; ++j) {
      double t = 0.0, *r = c[j];
      for (int k = 0; k < n; ++k) {
        t += p[k] * r[k];
      }
      q[j] = t;
    }
  }
  mm_destroy(n, c);
  return m;
}

double calc(int n) {
  n = n / 2 * 2;
  double **a = mm_gen(n, 1.0);
  double **b = mm_gen(n, 2.0);
  double **m = mm_mul(n, a, b);
  double result = m[n / 2][n / 2];
  mm_destroy(n, a);
  mm_destroy(n, b);
  mm_destroy(n, m);
  return result;
}

int main(int argc, char *argv[]) {
  int n = argc > 1 ? atoi(argv[1]) : 100;

  double left = calc(101);
  double right = -18.67;
  if (fabs(left - right) > 0.1) {
    fprintf(stderr, "%f != %f\n", left, right);
    exit(EXIT_FAILURE);
  }

  notify_with_pid("C/" COMPILER);
  double results = calc(n);
  notify("stop");

  printf("%f\n", results);
}
